examples/oft/README.md

```diff
@@ -374,6 +374,99 @@ By following these steps, you can focus more on creating innovative omnichain so
 
 <br></br>
 
+## Estimating `lzReceive` and `lzCompose` Gas Usage
+
+This guide explains how to use the `pnpm` commands to estimate gas usage for LayerZero's `lzReceive` and `lzCompose` functions. These commands wrap Foundry scripts for easier invocation and allow you to pass the required arguments dynamically.
+
+### Available Commands
+
+1. **`gas:lzReceive`**
+
+   This command profiles the `lzReceive` function for estimating gas usage across multiple runs.
+
+   ```json
+   "gas:lzReceive": "forge script scripts/GasProfiler.s.sol:GasProfilerScript --via-ir --sig 'run_lzReceive(string,address,uint32,address,uint32,address,bytes,uint256,uint256)'"
+   ```
+
+2. **`gas:lzCompose`**
+
+   This command profiles the `lzCompose` function for estimating gas usage across multiple runs.
+
+   ```json
+   "gas:lzCompose": "forge script scripts/GasProfiler.s.sol:GasProfilerScript --via-ir --sig 'run_lzCompose(string,address,uint32,address,uint32,address,address,bytes,uint256,uint256)'"
+   ```
+
+### Usage Examples
+
+#### `lzReceive`
+
+To estimate the gas for the `lzReceive` function:
+
+```bash
+pnpm gas:lzReceive
+  <rpcUrl> \
+  <endpointAddress> \
+  <srcEid> \
+  <sender> \
+  <dstEid> \
+  <receiver> \
+  <message> \
+  <msg.value> \
+  <numOfRuns>
```

Comment:
@DanL0 this is actually the correct amount.

After discussing with @tironiigor and @robingautier, I'm of the opinion that Developers should use the `ILayerZeroReceiver.lzReceive()` gasLimit in their `enforcedOptions` instead of the entire `endpoint.lzReceive()`.

:white_check_mark: Pros
Accurate Estimation: By using ILayerZeroReceiver.lzReceive, gas estimations are more precise as they account only for the execution costs specific to the receiver logic.
Reduced Complexity: Eliminates the need to factor in additional costs from the endpoint, simplifying the estimation process both off-chain and on-chain.

:warning: Cons
Requires Detailed Documentation: For teams implementing custom executors, it's essential to document and explain that enforcedOptions typically only include ILayerZeroReceiver.lzReceive costs, so they will need to price the base Endpoint execution costs for example _clearPayload into the total gasLimit amount.

In this case, the estimate is correct, as the Tenderly simulation you posted is:

![Screenshot 2024-12-04 at 11 45 55 AM](https://github.com/user-attachments/assets/e446b63f-dc70-4f25-9dea-2f81251f47a0)

The call after `_clearPayload` is the `ILayerZeroReceiver.lzReceive()`, which is only `19,051`.

Reply:
As for the arithmetic underflow, this is actually an RPC error.

One of the calls made is reverting with your RPC (likely rateLimits), since the `vm.createFork(rpcUrl)` sends a request to your RPC provider.

For example, my request was fine:

```bash
mattkrak ~/Documents/codebase/devtools/examples/oft [feat/gas-profiler] $ pnpm gas:lzReceive \
  https://polygon-amoy-bor-rpc.publicnode.com \
  0x6EDCE65403992e310A62460808c4b910D972f10f \
  40106 \
  0x669F19c625C1E82Cfc8b526779bDA6754fd01b38 \
  40267 \
  0x8AeE434e4186424fBe0E8b7a8DB72C4a4764EE55 \
  "0x0000000000000000000000000804a6e2798f42c7f3c97215ddf958d5500f8ec800000000000f4240" \
  0 \
  2 

> @layerzerolabs/oft-example@0.5.1 gas:lzReceive /Users/mattkrak/Documents/codebase/devtools/examples/oft
> forge script scripts/GasProfiler.s.sol:GasProfilerScript --via-ir --sig 'run_lzReceive(string,address,uint32,address,uint32,address,bytes,uint256,uint256)' "https://polygon-amoy-bor-rpc.publicnode.com" "0x6EDCE65403992e310A62460808c4b910D972f10f" "40106" "0x669F19c625C1E82Cfc8b526779bDA6754fd01b38" "40267" "0x8AeE434e4186424fBe0E8b7a8DB72C4a4764EE55" "0x0000000000000000000000000804a6e2798f42c7f3c97215ddf958d5500f8ec800000000000f4240" "0" "2"

[⠊] Compiling...
No files changed, compilation skipped
Script ran successfully.
Gas used: 93384

== Logs ==
  Starting gas profiling for lzReceive...
  Gas Usage Metrics:
  Average Gas Used: 21719
  Median Gas Used: 21719
  Maximum Gas Used: 21719
  Minimum Gas Used: 21719
  Successful Runs: 2
  ```

Reply:
interesting, without `--via-ir` the loop fails. Will take a deeper look, but for now I'd say including `--via-ir` in `package.json` is required.